### PR TITLE
arch/arm/samv7: Fix PWM operation for single channel mode

### PR DESCRIPTION
**Pull request by pkarashchenko**
## Summary
Fix PWM operation for single channel mode

The current SAMv7 PWM driver assumes that all PWM channels should
work in sync mode, but that is a partial case of a generic PWM
driver operation.

Start SAMv7 PWM channels in async mode. The sync mode should be
implemeted either using ioctl command or via a separate Kconfig
option.

## Impact
SAMv7 and IMXRT PWM

## Testing
Tested on SAME70-QMTECH board with logic analyzer.

### CODE REVIEW

File: arch/arm/src/samv7/sam_pwm.c
Comment (michallenc) [inline arch/arm/src/samv7/sam_pwm.c:471]: I am not sure about this. Lets say we configure only channels 2 and 3.

`int8_t index = info->channels[i].channel;` on line 456 gives as `index = 3` for channel 3. But then we would be out of `priv->channels[]` range with `priv->channels[3-1]` since only two channels are configured. There is no assurance all 4 channels will be configured.

I think the previous `info->channels[i].channel - 1` is sufficient here, all we need to do is just decrement the channel number used from application (from 1 to 4) so it would match register's offset (0 to 3).

The same goes for` pwm_set_output(dev, priv->channels[index - 1].channel, info->channels[i].duty);`.

I need to go now, I will be able to try the code in few hours.

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: Yes. I missed the index check. Will add it.

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: Range check added. Please try this patch

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: The `info->channels` hold a logical PWM channels and those could be mapped to different HW PWM channels. I mean we can have `CONFIG_PWM_NCHANNELS` set to 2 and having `CONFIG_SAMV7_PWM0_CH2` and `CONFIG_SAMV7_PWM0_CH3` enabled. So `info->channels` will result to `info->channels[0].channel == 1` and `info->channels[1].channel == 2` and should be mapped to HW channels 2 and 3. Do you agree?

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: I'm planning to add support of complementary output as well in the next couple of days

Comment (michallenc) [reply arch/arm/src/samv7/sam_pwm.c:471]: Should not the mapping be vice versa?. Or the question is what kind of numbers are we putting into `info->channels`. The number of channel (from 1 to 4 - CH0 = 1, CH1=2 etc...) or the position number (first configured channel equals 1, second equals 2 etc.)?

My understanding is that if we have config options `CONFIG_SAMV7_PWM0_CH0` to `CONFIG_SAMV7_PWM0_CH3` and we select `CONFIG_SAMV7_PWM0_CH2` and `CONFIG_SAMV7_PWM0_CH3` then channels' numbers put into `info->channels` should be 3 and 4 (we cannot use 0 so we need to index from 1 and this is third and fourth channel in SAM's PWM). It is better than mapping first and second channel from my point of view because user does not need to check what other channels are configured. I wasnt able to find any standard aproach in the documentation thought.

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: Yeah, I wasn't able to find any description as well. but my points are next:
1. The application (for example `examples/pwm`)  should be portable from board to board and across the chips (again this is purely my understanding) so it has to operate with "virtual" channels like it does now:
```
config EXAMPLES_PWM_CHANNEL1
	int "First PWM channel number"
	default 1
	range 1 6
	---help---
		The first PWM channel number.  Default: 1
```
2. Each specific board can implement PWM "virtual" channel based on specific HW details (map to HW channel 3 or 4 or even 567).

It is like with `/dev/timerX`, so `/dev/time1` does not mean that HW timer 1 is used.

But anyway let's move this discussion to the mailing thread.

Comment (michallenc) [reply arch/arm/src/samv7/sam_pwm.c:471]: I agree with your point about board to board (chip to chip) portability which is better with virtual channels. I tested your branch and managed to get both single channel and multiple channels output (using virtual channels in `example/pwm` app).

Yes, we can move the discussion regarding channel indexes to mailing list. I am now inclining towards virtual channels but we should document it properly somewhere.

Comment (xiaoxiang781216) [reply arch/arm/src/samv7/sam_pwm.c:471]: what's the final decision?

Comment (pkarashchenko) [reply arch/arm/src/samv7/sam_pwm.c:471]: I wrote this to mailing list, but no response till now. I think we need to make a clear description of how PWM multichannel works and update all the drivers accordingly.

Comment (michallenc) [reply arch/arm/src/samv7/sam_pwm.c:471]: I think we can merge this if there is no interest in mailing list. It would be great to standardized the PWM usage some day but I think it should not dealy this fix any further.

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**
